File: src/lib/api/plugins.ts
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import { createQuery } from "@tanstack/svelte-query";

import type {
  PluginId,
  PluginWithState,
  InspectorContext,
} from "./types/plugin";

import { queryClient } from "./client";

export const pluginsKey = {
  root: ["plugins"],
  list: ["plugins", "list"],
  specific: {
    asset: (pluginId: PluginId | null, asset: string | null) => [
      "plugins",
      "specific",
      pluginId,
      "asset",
      asset,
    ],
  },
};

// [REQUESTS] ------------------------------------------------------

export function getPluginsWithState() {
  return invoke<PluginWithState[]>("plugins_get_plugins");
}

export function getPluginProperties(pluginId: PluginId) {
  return invoke<object>("plugins_get_plugin_properties", { pluginId });
}

export function setPluginProperties(
  pluginId: PluginId,
  properties: object,
  partial: boolean = true,
) {
  return invoke<void>("plugins_set_plugin_properties", {
    pluginId,
    properties,
    partial,
  });
}

export function sendPluginMessage(context: InspectorContext, message: unknown) {
  return invoke<void>("plugins_send_plugin_message", {
    context,
    message,
  });
}

export function openPluginInspector(context: InspectorContext) {
  return invoke<void>("plugins_open_inspector", {
    context,
  });
}

export function closePluginInspector(context: InspectorContext) {
  return invoke<void>("plugins_close_inspector", {
    context,
  });
}

export async function stopPluginTask(pluginId: PluginId) {
  await invoke<void>("plugins_stop_plugin_task", {
    pluginId,
  });

  invalidatePluginsQuery();
}

export async function startPluginTask(pluginId: PluginId) {
  await invoke<void>("plugins_start_plugin_task", {
    pluginId,
  });

  invalidatePluginsQuery();
}

export async function restartPluginTask(pluginId: PluginId) {
  await invoke<void>("plugins_restart_plugin_task", {
    pluginId,
  });

  invalidatePluginsQuery();
}

export async function reloadPlugin(pluginId: PluginId) {
  await invoke<void>("plugins_reload_plugin", {
    pluginId,
  });

  invalidatePluginsQuery();
}

export async function installPlugin(file: File) {
  const data = await file.arrayBuffer();
  await installPluginBuffer(data);
}

export async function installPluginBuffer(data: ArrayBuffer) {
  await invoke<void>("plugins_install_plugin_manual", {
    data,
  });

  invalidatePluginsQuery();
}

export async function uninstallPlugin(pluginId: PluginId) {
  await invoke<void>("plugins_uninstall_plugin", {
    pluginId,
  });

  invalidatePluginsQuery();
}

// [QUERIES] ------------------------------------------------------

export function createPluginsQuery() {
  return createQuery({
    queryKey: pluginsKey.list,
    queryFn: getPluginsWithState,
  });
}

// [MUTATORS] ------------------------------------------------------

function invalidatePluginsQuery() {
  queryClient.invalidateQueries({
    queryKey: pluginsKey.root,
    exact: false,
  });
}

// [LISTENERS] ------------------------------------------------------

listen<PluginId>("plugins:loaded", ({ payload: plugin_id }) => {
  invalidatePluginsQuery();
});

listen<PluginId>("plugins:unloaded", ({ payload: plugin_id }) => {
  invalidatePluginsQuery();
});

listen<{ plugin_id: PluginId; state: string }>(
  "plugins:task_state_changed",
  ({ payload: { plugin_id, state } }) => {
    invalidatePluginsQuery();
  },
);

File: src-tauri/src/commands/plugins.rs
use std::sync::Arc;

use crate::{
    commands::CmdResult,
    database::JsonObject,
    events::InspectorContext,
    plugin::{
        PluginWithState, Plugins,
        install::{install_plugin_requirements, install_plugin_zip, remove_plugin_files},
        loader::{load_plugin_from_path, read_plugin_manifest_zip},
        manifest::PluginId,
    },
};
use tilepad_manifest::plugin::Manifest as PluginManifest;

use anyhow::Context;
use tauri::State;

#[tauri::command]
pub async fn plugins_install_plugin_manual(
    plugins: State<'_, Arc<Plugins>>,
    data: Vec<u8>,
) -> CmdResult<()> {
    // Read the plugin manifest from within the zip file
    let manifest = read_plugin_manifest_zip(&data).await?;

    // Determine plugin install directory
    let plugin_id = &manifest.plugin.id;
    let path = plugins.user_path().join(&plugin_id.0);

    // Unload the plugin
    plugins.unload_plugin(plugin_id).await;

    // Cleanup old files
    remove_plugin_files(&path).await?;

    // Install the plugin zip file
    install_plugin_zip(&data, &path).await?;

    // Install plugin runtime if one is required
    install_plugin_requirements(&manifest, &plugins.runtimes_path()).await?;

    // Load the plugin
    let plugin = load_plugin_from_path(&path)
        .await
        .context("failed to load plugin")?;

    plugins.load_plugin(plugin).await;

    Ok(())
}

#[tauri::command]
pub async fn plugins_uninstall_plugin(
    plugins: State<'_, Arc<Plugins>>,
    plugin_id: PluginId,
) -> CmdResult<()> {
    // Determine plugin install directory
    let path = plugins.user_path().join(&plugin_id.0);

    // Unload the plugin
    plugins.unload_plugin(&plugin_id).await;

    // Cleanup old files
    remove_plugin_files(&path).await?;

    Ok(())
}

#[tauri::command]
pub async fn plugins_send_plugin_message(
    plugins: State<'_, Arc<Plugins>>,
    context: InspectorContext,
    message: serde_json::Value,
) -> CmdResult<()> {
    plugins.handle_send_message(context, message).await?;

    Ok(())
}

#[tauri::command]
pub async fn plugins_open_inspector(
    plugins: State<'_, Arc<Plugins>>,
    context: InspectorContext,
) -> CmdResult<()> {
    plugins.open_inspector(context);
    Ok(())
}

#[tauri::command]
pub async fn plugins_close_inspector(
    plugins: State<'_, Arc<Plugins>>,
    context: InspectorContext,
) -> CmdResult<()> {
    plugins.close_inspector(context);
    Ok(())
}

#[tauri::command]
pub async fn plugins_get_plugin_properties(
    plugins: State<'_, Arc<Plugins>>,
    plugin_id: PluginId,
) -> CmdResult<JsonObject> {
    let result = plugins.get_plugin_properties(plugin_id).await?;
    Ok(result)
}

#[tauri::command]
pub async fn plugins_set_plugin_properties(
    plugins: State<'_, Arc<Plugins>>,
    plugin_id: PluginId,
    properties: JsonObject,
    partial: bool,
) -> CmdResult<()> {
    plugins
        .set_plugin_properties(plugin_id, properties, partial)
        .await?;
    Ok(())
}

#[tauri::command]
pub fn plugins_get_plugins(plugins: State<'_, Arc<Plugins>>) -> Vec<PluginWithState> {
    plugins.get_plugins_with_state()
}

#[tauri::command]
pub async fn plugins_stop_plugin_task(
    plugins: State<'_, Arc<Plugins>>,
    plugin_id: PluginId,
) -> CmdResult<()> {
    plugins.stop_task(&plugin_id).await;
    Ok(())
}

#[tauri::command]
pub async fn plugins_start_plugin_task(
    plugins: State<'_, Arc<Plugins>>,
    plugin_id: PluginId,
) -> CmdResult<()> {
    let plugin = plugins.get_plugin(&plugin_id).context("plugin not found")?;

    plugins
        .start_task(plugin_id, plugin.path.clone(), &plugin.manifest)
        .await;

    Ok(())
}

#[tauri::command]
pub async fn plugins_restart_plugin_task(
    plugins: State<'_, Arc<Plugins>>,
    plugin_id: PluginId,
) -> CmdResult<()> {
    let plugin = plugins.get_plugin(&plugin_id).context("plugin not found")?;

    plugins
        .restart_task(plugin_id, plugin.path.clone(), &plugin.manifest)
        .await;

    Ok(())
}

#[tauri::command]
pub async fn plugins_reload_plugin(
    plugins: State<'_, Arc<Plugins>>,
    plugin_id: PluginId,
) -> CmdResult<()> {
    // Unload the plugin
    let plugin = plugins
        .unload_plugin(&plugin_id)
        .await
        .context("plugin was never loaded")?;

    // Load the new plugin from the same path
    let new_plugin = load_plugin_from_path(&plugin.path).await?;

    // Load the new plugin into the plugins registry
    plugins.load_plugin(new_plugin).await;
    Ok(())
}

#[tauri::command]
pub fn plugins_parse_manifest(manifest: String) -> CmdResult<PluginManifest> {
    let manifest: PluginManifest = PluginManifest::parse(&manifest)?;
    Ok(manifest)
}
